### PR TITLE
[fix][flink] Fix Iceberg compatible metadata cannot be read when data file format is parquet

### PR DESCRIPTION
**Pull request by tsreaper**
### Purpose

Currently there is not IT cases for Paimon Iceberg compatibility in Flink. This PR adds IT cases and fixes some issues with parquet format.

### Tests

IT cases.

### API and Format

No format changes.

### Documentation

Document for the whole Iceberg compatibility feature is also added.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+                    .intType()
+                    .defaultValue(10);
+
+    public static final ConfigOption<Integer> COMPACT_MAX_FILE_NUM =
+            ConfigOptions.key("metadata.iceberg.compaction.max.file-num")
+                    .intType()
+                    .defaultValue(50);
+
+    /** Where to store Iceberg metadata. */
+    public enum StorageType {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:51]: extends DescribedEnum

Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:51]: Values can be: NONE, TABLE_PATH, HADOOP_CATALOG?

Comment (tsreaper) [reply paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:51]: Why `NONE`? This value should not be set if users don't need Iceberg compatibility.

Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:51]: Or name it `DISABLED`

Comment (JingsongLi) [reply paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:51]: Generally speaking, using ENUM already has a state inside, and disable should also be one of its states

File: paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java [hunk 2]
```diff
@@ -48,11 +48,22 @@ public class IcebergOptions {
...
-        // Store Iceberg metadata in a separate directory.
-        // This directory can be specified as the Iceberg warehouse directory.
-        ICEBERG_WAREHOUSE
+    public enum StorageType implements DescribedEnum {
+        DISABLED("Disable Iceberg compatibility support."),
+        TABLE_LOCATION("Store Iceberg metadata with each table."),
+        HADOOP_CATALOG(
+                "Store Iceberg metadata in a separate directory. "
+                        + "This directory can be specified as the warehouse directory of an Iceberg Hadoop catalog.");
+
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:57]: You can add a value and implement `toString`. For the paimon style, the value should be `table-location` and `hadoop-catalog`.

File: paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                    .enumType(StorageType.class)
+                    .defaultValue(StorageType.DISABLED)
+                    .withDescription(
+                            "When set, produce Iceberg metadata after a snapshot is committed, "
+                                    + "so that Iceberg readers can read Paimon's raw data files.");
+
+    public static final ConfigOption<Integer> COMPACT_MIN_FILE_NUM =
+            ConfigOptions.key("metadata.iceberg.compaction.min.file-num")
+                    .intType()
+                    .defaultValue(10);
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:43]: add withDescription?

File: paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+    public static final ConfigOption<Integer> COMPACT_MAX_FILE_NUM =
+            ConfigOptions.key("metadata.iceberg.compaction.max.file-num")
+                    .intType()
+                    .defaultValue(50);
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:48]: add withDescription?

File: paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+    /** Where to store Iceberg metadata. */
+    public enum StorageType implements DescribedEnum {
+        DISABLED("Disable Iceberg compatibility support."),
+        TABLE_LOCATION("Store Iceberg metadata with each table."),
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/IcebergOptions.java:53]: Store Iceberg metadata in each table's location.

File: paimon-core/src/main/java/org/apache/paimon/iceberg/metadata/IcebergSchema.java [hunk 6]
```diff
@@ -50,11 +51,25 @@ public class IcebergSchema {
     @JsonProperty(FIELD_FIELDS)
     private final List<IcebergDataField> fields;
 
-    public IcebergSchema(TableSchema tableSchema) {
-        this(
+    public static IcebergSchema create(TableSchema tableSchema) {
+        int bias;
+        if (new CoreOptions(tableSchema.options()).formatType().equals("parquet")) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/metadata/IcebergSchema.java:56]: use CoreOptions.FILE_FORMAT_PARQUET

File: paimon-core/src/main/java/org/apache/paimon/iceberg/metadata/IcebergSchema.java [hunk 7]
```diff
@@ -50,11 +51,25 @@ public class IcebergSchema {
...
+            bias = 0;
+        }
+
+        return new IcebergSchema(
                 (int) tableSchema.id(),
                 tableSchema.fields().stream()
-                        .map(IcebergDataField::new)
+                        .map(f -> new IcebergDataField(f, bias))
                         .collect(Collectors.toList()));
     }
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/iceberg/metadata/IcebergSchema.java:74]: add "private static final String STRUCT = "struct";" for line 77

File: paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java [hunk 8]
```diff
@@ -165,7 +166,8 @@ protected List<CommitCallback> createCommitCallbacks(String commitUser) {
         List<CommitCallback> callbacks = super.createCommitCallbacks(commitUser);
         CoreOptions options = coreOptions();
 
-        if (options.metadataIcebergCompatible()) {
+        if (options.toConfiguration().get(IcebergOptions.METADATA_ICEBERG_STORAGE)
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/AppendOnlyFileStoreTable.java:169]: Move this to AbstractFileStoreTable for AppendOnlyFileStoreTable and PrimaryKeyFileStoreTable?

File: paimon-core/src/main/java/org/apache/paimon/table/PrimaryKeyFileStoreTable.java [hunk 9]
```diff
@@ -185,7 +186,8 @@ protected List<CommitCallback> createCommitCallbacks(String commitUser) {
         List<CommitCallback> callbacks = super.createCommitCallbacks(commitUser);
         CoreOptions options = coreOptions();
 
-        if (options.metadataIcebergCompatible()) {
+        if (options.toConfiguration().get(IcebergOptions.METADATA_ICEBERG_STORAGE)
+                != IcebergOptions.StorageType.DISABLED) {
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/table/PrimaryKeyFileStoreTable.java:190]: Move this to AbstractFileStoreTable for AppendOnlyFileStoreTable and PrimaryKeyFileStoreTable?

File: paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/iceberg/FlinkIcebergITCaseBase.java [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+import org.apache.flink.util.CloseableIterator;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** IT cases for Paimon Iceberg compatibility. */
```
Comment (wwj6591812) [inline paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/iceberg/FlinkIcebergITCaseBase.java:35]: Base IT cases for Paimon Iceberg compatibility.

**Review by JingsongLi — APPROVED** [review]
Looks good to me!

**Pull request merged**